src/Rules/Library.hs

```diff
@@ -127,10 +109,221 @@ cObjects context = do
         then objs
         else filter ((`notElem` ["Evac_thr", "Scav_thr"]) . takeBaseName) objs
 
+-- | Return extra object files needed to build the given library context. The
+--   resulting list is non empty only when the package from the 'Context' is
+--   /integer-gmp/.
 extraObjects :: Context -> Action [FilePath]
 extraObjects context
     | package context == integerGmp = do
         gmpPath <- gmpBuildPath
         need [gmpPath -/- gmpLibraryH]
         map unifyPath <$> getDirectoryFiles "" [gmpPath -/- gmpObjectsDir -/- "*.o"]
     | otherwise         = return []
+
+-- | Return all the object files to be put into the library
+--   we're building for the given 'Context'.
+libraryObjects :: Context -> Action [FilePath]
+libraryObjects context@Context{..} = do
+    hsObjs   <- hsObjects    context
+    noHsObjs <- nonHsObjects context
+
+    -- This will create split objects if required (we don't track them
+    -- explicitly as this would needlessly bloat the Shake database).
+    need $ noHsObjs ++ hsObjs
+
+    split <- interpretInContext context =<< splitObjects <$> flavour
+    let getSplitObjs = concatForM hsObjs $ \obj -> do
+            let dir = dropExtension obj ++ "_" ++ osuf way ++ "_split"
+            contents <- liftIO $ IO.getDirectoryContents dir
+            return . map (dir -/-) $ filter (not . all (== '.')) contents
+
+    (noHsObjs ++) <$> if split then getSplitObjs else return hsObjs
+
+-- * Library paths types and parsers
+
+-- | > libHS<pkg name>-<pkg version>[_<way suffix>].a
+data LibA = LibA String [Integer] Way
+  deriving (Eq, Show)
+
+-- | > <so or dylib>
+data DynLibExt = So | Dylib
+  deriving (Eq, Show)
+
+-- | > libHS<pkg name>-<pkg version>-ghc<ghc version>[_<way suffix>].<so or dylib>
+data LibDyn = LibDyn String [Integer] Way DynLibExt
+  deriving (Eq, Show)
+
+-- | > HS<pkg name>-<pkg version>[_<way suffix>].o
+data LibGhci = LibGhci String [Integer] Way
+  deriving (Eq, Show)
+
+-- | A path of the form
+--
+--   > <build root>/stage<N>/<path/to/pkg/from/ghc/root>/build/<something>
+--
+--   where @something@ describes a library to be build for the given package.
+--
+--   @a@, which represents that @something@, is instantiated as 'LibA', 'LibDyn'
+--   and 'LibGhci' successively in this module, depending on the type of library
+--   we're giving the build rules for.
+data BuildPath a
```

Comment:
Presumably someday this will become a generic piece of infastructure and will live elsewhere, but no need to move things right now. I'm just commenting that this is potentially way to valuable to be hidden in `Rules.Libarary` :)

Reply:
Right, I find this rather useful. I too like the parser-based approach combined with things like `BuildPath`, I didn't move it to a dedicated module and what not because for now we're just using it there, but I was curious to see whether we would like to do something similar "everywhere we can". Glad to see the positive feedback.